### ISSUE TITLE
WarmupDecayLR.params.total_num_steps - total or per gpu?

### ISSUE THREAD

**Issue by stas00** [microsoft/DeepSpeed#633]
We are having a bit of hard time getting `total_num_steps` to pass to `WarmupDecayLR` at init time - it's a bit too early for the logic as these points are configured once ddp/ds has started - we found a workaround, but it doesn't take into the account the number of gpus.

I wanted to check that `WarmupDecayLR.params.total_num_steps` expects the total for the whole world and not per gpu.

Currently the doc says:

```
class WarmupDecayLR(WarmupLR):
[...]
            total_num_steps (int): total number of training steps
```
so it'd be awesome to disambiguate the context.

Thank you!

**Comment by tjruwase** [comment]
I find it easier to think of training step in terms of effective batch size. So while this implies world size, it also incorporates gradient accumulation steps. I guess another way is to say training steps == optimizer steps. Does this help? I will update doc shortly.

**Comment by stas00** [comment]
I'm not sure it helps. While it can be overriden by users with `--max_steps`, normally it's derived automatically at run time based on the dataset length, and gradient_accumulation_steps, to be exact:

```
            num_update_steps_per_epoch = len(train_dataloader) // self.args.gradient_accumulation_steps
            num_update_steps_per_epoch = max(num_update_steps_per_epoch, 1)
            return math.ceil(self.args.num_train_epochs * num_update_steps_per_epoch)
```

So the modified question with this context - do we need to divide this by number of gpus or not? 

My guess is that since you said "Effective batch size" then yes it should be divided by n_gpus, correct?

**Comment by tjruwase** [comment]
Besides the chicken and egg problem, I am a bit confused about that code snippet for computing training steps. It seems to be independent of effective batch size, or am I missing something? 

Also is len(train_dataloader) per gpu or for the world size?

**Comment by stas00** [comment]
When I launch the program with:
```
deepspeed  ./finetune_trainer.py ...
```
It does know the world size early on according to the logs:
```
[2021-01-05 13:30:20,634] [INFO] [launch.py:100:main] dist_world_size=2
```
but it won't `torch.distributed.init_process_group()` until after `deepspeed.initialize` is initialized and it's too late for our needs.

I guess I could have hacked around this if I could get the world size by simply measuring the len of the whole dataset and dividing it by the world size, which would give a roughly pretty close number, but again I can't call `dist.get_world_size()` at that stage.

**Comment by stas00** [comment]
Perhaps the correct solution is that there should be 2 init functions:

1. `deepspeed.init_dist` which the user should call as soon as possible - first thing in the script as soon as args have been parsed
2. `deepspeed.initialize` - the rest of it.

**Comment by stas00** [comment]
Would it be possible to look at this issue, please?

**Comment by jeffra** [comment]
If you're needing the world size _before_ the deepspeed.initialize call then yes I think that's perfect motivation for calling `deepspeed.init_distributed()` sometime earlier. This will also help in other scenarios where torch.distributed needs to be used before deepspeed.initialize is invoked (e.g., model parallelism).

**Comment by stas00** [comment]
oh, you have recently added just that - fantastic - I will give it a try and report back. Thank you, @jeffra

**Comment by stas00** [comment]
It worked beautifully and greatly simplified our code - thank you, @jeffra 

It'd be awesome to document it at https://www.deepspeed.ai/getting-started/#writing-deepspeed-models perhaps?

I've proposed this doc to reflect this important function https://github.com/microsoft/DeepSpeed/pull/644
